### PR TITLE
Fix CMake files in libcudf C++ examples to use existing libcudf build if present

### PR DESCRIPTION
**Pull request by mhaseeb123**
## Description
This PR fixes the CMake artifacts for libcudf examples and includes CI updates to create executable `libcudf-example` conda package to run from CI 

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/examples/build.sh
Comment (bdice) [reply cpp/examples/build.sh:24]: I would like to update our CI testing to run the example binaries as part of the C++ tests. The CI system where this is built won't have a GPU, and I would like the example package to be runnable on all RAPIDS-supported hardware for CI reasons.

File: cpp/examples/fetch_dependencies.cmake
Comment (bdice) [inline cpp/examples/fetch_dependencies.cmake:29]: I think CPMFindPackage can search for a matching local package. Is there something missing here that we need to make it discoverable? Do we need to add something like `VERSION 24.06.00` to the CPMFindPackage call?

Comment (mhaseeb123) [reply cpp/examples/fetch_dependencies.cmake:29]: Good suggestion, I think CPM's default behavior is to download first but I think I can get around it and have it look locally first.

File: cpp/examples/build.sh
Comment (mhaseeb123) [reply cpp/examples/build.sh:13]: Doubleminded between this and @davidwendt's comment. Should I hardcode to `NATIVE' or call `rapids_cuda_set_architectures(RAPIDS)`

Comment (bdice) [reply cpp/examples/build.sh:13]: I would use rapids-cmake’s official list and not NATIVE. We want the libcudf-example binaries to be runnable on all supported arches.

File: cpp/examples/build.sh
Comment (mhaseeb123) [reply cpp/examples/build.sh]: All done

File: cpp/examples/build.sh
Comment (bdice) [inline cpp/examples/build.sh:23]: Since this file has a hardcoded version number, you have to update `ci/release/update-version.sh` so it stays in sync. You probably need to edit the line that is currently in there, since `fetch_dependencies.cmake` no longer has the version number:

https://github.com/rapidsai/cudf/blob/79d2dba141e0cbc464810a0b214d2c714ebe1a18/ci/release/update-version.sh#L72-L73

File: cpp/examples/set_cuda_architecture.cmake
Comment (bdice) [inline cpp/examples/set_cuda_architecture.cmake:2]: New file should just be copyright 2024.
```suggestion
# Copyright (c) 2024, NVIDIA CORPORATION.
```

**Review by bdice — CHANGES_REQUESTED** [review]
Great. This is getting closer to what I expect.

Do you think we should also run the binaries from `libcudf-example` in the C++ tests job in CI?

The steps to do that:

Add `libcudf-example` to https://github.com/rapidsai/cudf/blob/79d2dba141e0cbc464810a0b214d2c714ebe1a18/ci/test_cpp_common.sh#L32-L34

Then add a section like this to the `test_cpp.sh` job that calls a new script called `run_cudf_examples.sh`. Track the error/exit code. https://github.com/rapidsai/cudf/blob/79d2dba141e0cbc464810a0b214d2c714ebe1a18/ci/test_cpp.sh#L16-L18

Test that this fails in CI by pushing a change that compiles to an example binary but exits with a nonzero code.

**Comment by davidwendt** [comment]
> Do you think we should also run the binaries from `libcudf-example` in the C++ tests job in CI?

Just to be clear. This would just run the example to make sure it did not crash? 
It would not actually verify the example produces the correct result, right?
I suppose there is value is checking for a crash only.
Perhaps it could run through compute-sanitizer.
This may be out of scope for this PR.

**Comment by bdice** [comment]
Yes, this is just verify the example binaries (1) were built, (2) were packaged, and (3) return 0 / do not crash. The binaries could be run under sanitizers in the other test script but that is lower priority to me.

**Comment by bdice** [comment]
We could run the examples in CI in another PR but I don’t want to lose track of this. I’d prefer it in this PR since the changes should be relatively small and are somewhat related to the other build fixes here.

**Comment by harrism** [comment]
@mhaseeb123 would you mind making your PR title a little more specific? The title goes in CHANGELOG.md when we release so specific titles help make this file useful.

File: cpp/examples/build.sh
Comment (mhaseeb123) [reply cpp/examples/build.sh]: Put the `CUDF_TAG` back in `fetch_dependencies.cmake` and imported it in `set_cuda_architecture.cmake` for consistency.

**Comment by mhaseeb123** [comment]
> @mhaseeb123 would you mind making your PR title a little more specific? The title goes in CHANGELOG.md when we release so specific titles help make this file useful.

Updated the PR title.

File: conda/recipes/libcudf/install_libcudf_example.sh
Comment (bdice) [inline conda/recipes/libcudf/install_libcudf_example.sh]: Is this logic to store the old value necessary? I think this environment variable would only affect this script's environment, and not other scripts.

File: cpp/examples/basic/CMakeLists.txt
Comment (bdice) [inline cpp/examples/basic/CMakeLists.txt:24]: This seems like surprising behavior to install if an environment variable is present, especially since this isn't a RAPIDS-specific environment variable. We shouldn't need to set this variable explicitly, since CMake is supposed to read it from the environment. https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html

What happens if you remove this conditional? Do we need to specify something like `install(... COMPONENT examples EXCLUDE_FROM_ALL)`?

See also: https://discourse.cmake.org/t/understanding-the-cmake-component-keyword-in-the-install-command/971

File: cpp/examples/build.sh
Comment (bdice) [inline cpp/examples/build.sh:32]: I think you're supposed to set `CMAKE_INSTALL_PREFIX` (the environment variable in our case, since we want it to be controlled by `install_cudf_examples.sh` rather than this script) when CMake runs its configure step, then you shouldn't need the `--prefix` here. I'm still learning this part of CMake so I'd be happy to have corrections.

File: cpp/examples/fetch_dependencies.cmake
Comment (bdice) [inline cpp/examples/fetch_dependencies.cmake:23]: Do we need this or is it temporary for local testing? To my knowledge, you shouldn't need this if you have a local package that meets the version requirements in a place that CMake can find it.

**Review by bdice — COMMENTED** [review]
I have a few comments for now -- I think we'll need some additional CMake eyes from @robertmaynard, @KyleFromNVIDIA, or @vyasr before merging.

tl;dr this PR is trying to fix the libcudf examples and make the `libcudf-example` conda package have runnable binaries that we can execute in CI tests.

File: conda/recipes/libcudf/install_libcudf_example.sh
Comment (mhaseeb123) [reply conda/recipes/libcudf/install_libcudf_example.sh]: Nah, not necessary, I was trying to locally experiment which got committed.

File: cpp/examples/build.sh
Comment (mhaseeb123) [reply cpp/examples/build.sh:32]: Still cleaning these up but the `--prefix` option in `cmake --install` is pretty much useless and does not take any precedence over the initially set `CMAKE_INSTALL_PREFIX`.

File: cpp/examples/fetch_dependencies.cmake
Comment (mhaseeb123) [reply cpp/examples/fetch_dependencies.cmake:23]: Yeah not needed. Will clean up soon

File: conda/recipes/libcudf/install_libcudf_example.sh
Comment (mhaseeb123) [reply conda/recipes/libcudf/install_libcudf_example.sh]: Not necessary at all. Some local testing artifacts seeped through here.

File: cpp/examples/basic/CMakeLists.txt
Comment (mhaseeb123) [reply cpp/examples/basic/CMakeLists.txt:24]: Agreed with this. We need a variable (currently I am using the CMAKE_INSTALL_PREFIX itself to test out the CMake's shenanigans) to control installation and have it to set `CMAKE_INSTALL_PREFIX` too in the initial `CMake` command. If not set at this stage, it has to be forced before `install` in CMakeLists.txt using some check

File: cpp/examples/build.sh
Comment (robertmaynard) [reply cpp/examples/build.sh]: The install rules in this example are written in a way that is fundamentally incompatible with `cmake --install --prefix`.

The cmake install command docs outline the issue:
```
DESTINATION <dir>
Specify the directory on disk to which a file will be installed. <dir> should be a relative path. An absolute path is allowed, but not recommended.

When a relative path is given it is interpreted relative to the value of the [CMAKE_INSTALL_PREFIX](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html#variable:CMAKE_INSTALL_PREFIX) variable. The prefix can be relocated at install time using the DESTDIR mechanism explained in the [CMAKE_INSTALL_PREFIX](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html#variable:CMAKE_INSTALL_PREFIX) variable documentation.

As absolute paths do not work with the cmake --install command's [--prefix](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake--install-0) option, or with the [cpack](https://cmake.org/cmake/help/latest/manual/cpack.1.html#manual:cpack(1)) installer generators, it is strongly recommended to use relative paths throughout for best support by package maintainers. In particular, there is no need to make paths absolute by prepending [CMAKE_INSTALL_PREFIX](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html#variable:CMAKE_INSTALL_PREFIX); this prefix is used by default if the DESTINATION is a relative path. 
```

In all the install rules you are using `${CMAKE_INSTALL_PREFIX}` which is not needed since all relative paths are evaluatd based off of it, and secondly creating absolute paths which can't be re-rooted by `--prefix`.

File: cpp/examples/basic/CMakeLists.txt
Comment (robertmaynard) [inline cpp/examples/basic/CMakeLists.txt:24]: We should just require cmake 3.29 if you want to support CMAKE_INSTALL_PREFIX from the env, since that is the version where the variable is supported. 

Otherwise you are doing non standard behavior that isn't expected by CMake users.


Looking over the entire code, I expect that the following is going to simplify the logic and resolve your issues;

1. Yank all usages of `${CMAKE_INSTALL_PREFIX}` from the install commands. It is not needed and is an anti-pattern that is break `cmake --install --prefix`.
2. Yank all uages of trying to specify `CMAKE_INSTALL_PREFIX` from the env in the CMakeLists.txt
3. Yank all setting of the CMAKE_INSTALL_PREFIX in the conda recipe scripts.
4. Add something like a `--install` option to `cpp/examples/build.sh` which will call `cmake --install --prefix` using some logic where the install prefix location is deduced via something like: `INSTALL_PREFIX=${INSTALL_PREFIX:=${PREFIX:=${CONDA_PREFIX}}}`

**Review by robertmaynard — CHANGES_REQUESTED** [review]

File: cpp/examples/set_cuda_architecture.cmake
Comment (KyleFromNVIDIA) [inline cpp/examples/set_cuda_architecture.cmake:20]: Instead of `grep`ping and `sed`ing a CMake file, any data should be stored in either a `.cmake` file that sets variables or a JSON file that can be read with `string(JSON)`. Then `fetch_dependencies.cmake` and `set_cuda_architecture.cmake` can both read from this single source of truth.

**Review by KyleFromNVIDIA — DISMISSED** [review]

File: conda/recipes/libcudf/meta.yaml
Comment (bdice) [inline conda/recipes/libcudf/meta.yaml:198]: Match the naming of `libcudf-tests`, which says "libcudf test & benchmark executables"
```suggestion
      summary: libcudf example executables
```

File: cpp/examples/set_cuda_architecture.cmake
Comment (mhaseeb123) [reply cpp/examples/set_cuda_architecture.cmake:20]: The reason to keep `CUDF_TAG` in `fetch_dependencies.cmake was to not break any other CI artifacts reading from here like: https://github.com/rapidsai/cudf/blob/79d2dba141e0cbc464810a0b214d2c714ebe1a18/ci/release/update-version.sh#L72-L73. Let me just avoid all this everywhere.

File: cpp/examples/basic/CMakeLists.txt
Comment (mhaseeb123) [reply cpp/examples/basic/CMakeLists.txt:24]: Thanks for the suggestions Robert. Incorporating these changes in the next commit.

File: conda/recipes/libcudf/meta.yaml
Comment (mhaseeb123) [reply conda/recipes/libcudf/meta.yaml]: All done!

File: cpp/examples/set_cuda_architecture.cmake
Comment (mhaseeb123) [reply cpp/examples/set_cuda_architecture.cmake]: All done!

File: cpp/examples/basic/CMakeLists.txt
Comment (mhaseeb123) [reply cpp/examples/basic/CMakeLists.txt]: All done!

File: cpp/examples/build.sh
Comment (bdice) [inline cpp/examples/build.sh:47]: Where does `INSTALL_PREFIX` come from? Do we need all of these environment variables? Could we separate this onto another line?

I'd propose something like:
```suggestion
    if [ "$CONDA_BUILD" == "1" ]; then
      INSTALL_PREFIX="${INSTALL_PREFIX:-${PREFIX}}"
    fi
    INSTALL_PREFIX="${INSTALL_PREFIX:-${example_dir}/install}"
    cmake --install ${build_dir} --prefix ${INSTALL_PREFIX}
```

File: cpp/examples/fetch_dependencies.cmake
Comment (bdice) [inline cpp/examples/fetch_dependencies.cmake:28]: Do we still need this `HINTS`? I think that `latest` is a devcontainer-specific name. We probably don't want that.

**Review by bdice — COMMENTED** [review]
Nice. This is vastly improved. I have a few comments, but we're getting closer.

File: cpp/examples/fetch_dependencies.cmake
Comment (mhaseeb123) [reply cpp/examples/fetch_dependencies.cmake:28]: We do need `HINTS` to tell CMake where to `find_package()` libcudf installation. I think we can use both `${cudf_ROOT}:${cudf_ROOT}/latest' as HINTS so it works with devcontainer and otherwise. Though, it would fall back to installing from `GIT_REPOSITORY` if it can't find the package.

File: cpp/examples/build.sh
Comment (mhaseeb123) [reply cpp/examples/build.sh:47]: We don't need all these, just any one of them (in the preference order) set in the environment. I am partial to removing `$CONDA_PREFIX` from this chain since it adds a possibly unrelated path. Afaik `$INSTALL_PREFIX` is not used by any of the automated tests so it allows users to override the prefix, whereas its first fallback on `$PREFIX` is used by CI to correctly run stuff.

Comment (mhaseeb123) [reply cpp/examples/build.sh:47]: Afaik `$INSTALL_PREFIX` is not being used by CI so it can be used by user to provide a superseding path. 

Only of the variables in the chain need to be set.

If `$INSTALL_PREFIX` isn't set, we should set to `${PREFIX}` if running from CI (is `$CONDA_BUILD="1"` check good for this?), else to the local `${example_dir}/install` like you suggested.

Comment (bdice) [reply cpp/examples/build.sh:47]: > Small question, is the `CONDA_BUILD` set to `1` by the CI build? I couldn't find this env var in the recent CI job.

Yes. It's set by `conda-build`.

Reference: https://docs.conda.io/projects/conda-build/en/stable/user-guide/environment-variables.html#environment-variables-set-during-the-build-process

Comment (bdice) [reply cpp/examples/build.sh:47]: > Afaik `$INSTALL_PREFIX` is not being used by CI so it can be used by user to provide a superseding path.
> 
> Only of the variables in the chain need to be set.
> 
> If `$INSTALL_PREFIX` isn't set, we should set to `${PREFIX}` if running from CI (is `$CONDA_BUILD="1"` check good for this?), else to the local `${example_dir}/install` like you suggested.

Yes, you're reading my intentions correctly! I hoped it would be self-explanatory, and you got it right. 👍 Let's apply the suggestion above.

File: cpp/examples/fetch_dependencies.cmake
Comment (bdice) [reply cpp/examples/fetch_dependencies.cmake:28]: Okay, this is fine with me.

File: cpp/examples/build.sh
Comment (mhaseeb123) [reply cpp/examples/build.sh:47]: Perfect. Let me do that!

File: cpp/examples/fetch_dependencies.cmake
Comment (mhaseeb123) [reply cpp/examples/fetch_dependencies.cmake]: All done!

File: cpp/examples/basic/CMakeLists.txt
Comment (mhaseeb123) [reply cpp/examples/basic/CMakeLists.txt]: All done!

**Comment by mhaseeb123** [comment]
Weird CI build failure. Retriggering run /ok to test

File: cpp/examples/fetch_dependencies.cmake
Comment (robertmaynard) [reply cpp/examples/fetch_dependencies.cmake]: `HINTS` shouldn't be used. `HINTS` are meant to be locations that are deduced by system introspection. What we have here is a guess based on a layout that occurs only under certain situations.  What you want is to pass this in `PATHS`, as that has significantly lower search priority. 


This does raise the followup question of why aren't builds in devcontainers passing a `-Dcudf_ROOT` or `-Dcudf_DIR`  argument when building the examples. IIRC they are already doing things like this for other projects.

File: cpp/examples/strings/CMakeLists.txt
Comment (robertmaynard) [inline cpp/examples/strings/CMakeLists.txt:42]:
```suggestion
  DESTINATION
```

**Review by robertmaynard — CHANGES_REQUESTED** [review]

**Review by bdice — APPROVED** [review]
This seems fine to me. I think it'd be good to get approval from @robertmaynard / @KyleFromNVIDIA / @davidwendt since they all requested changes.

File: cpp/examples/strings/CMakeLists.txt
Comment (robertmaynard) [reply cpp/examples/strings/CMakeLists.txt]: Thank you

**Review by robertmaynard — APPROVED** [review]

**Review by AyodeAwe — APPROVED** [review]

**Pull request merged**
